src/sdk/pynni/nni/compression/torch/pruning/amc/channel_pruning_env.py

```diff
@@ -0,0 +1,497 @@
+
+        masks = self.channel_prune_masker.calc_mask(sparsity=1-preserve_ratio, wrapper=op, preserve_idx=preserve_idx)
+        m = masks['weight_mask'].cpu().data
+        if type(op.module) == nn.Conv2d:
+            d_prime = (m.sum((0, 2, 3)) > 0).sum().item()
+            preserve_idx = np.nonzero((m.sum((0, 2, 3)) > 0).numpy())[0]
+        else:
+            assert type(op.module) == nn.Linear
+            d_prime = (m.sum(1) > 0).sum().item()
+            preserve_idx = np.nonzero((m.sum(1) > 0).numpy())[0]
+
+        op.weight_mask = masks['weight_mask']
+        if hasattr(op.module, 'bias') and op.module.bias is not None and 'bias_mask' in masks:
+            op.bias_mask = masks['bias_mask']
+
+        action = (m == 1).sum().item() / m.numel()
+        return action, d_prime, preserve_idx
+
+    def export_model(self):
+        while True:
+            self.export_layer(self.prunable_idx[self.cur_ind])
+            if self._is_final_layer():
+                break
+            self.cur_ind += 1
+
+    def export_layer(self, op_idx):
+        m_list = list(self.model.modules())
+        op = m_list[op_idx]
+        assert type(op) == PrunerModuleWrapper
+        w = op.module.weight.cpu().data
+        m = op.weight_mask.cpu().data
+        if type(op.module) == nn.Linear:
+            w = w.unsqueeze(-1).unsqueeze(-1)
+            m = m.unsqueeze(-1).unsqueeze(-1)
+
+        d_prime = (m.sum((0, 2, 3)) > 0).sum().item()
+        preserve_idx = np.nonzero((m.sum((0, 2, 3)) > 0).numpy())[0]
+        assert d_prime <= w.size(1)
+
+        if d_prime == w.size(1):
+            return
+
+        mask = np.zeros(w.size(1), bool)
+        mask[preserve_idx] = True
+        rec_weight = torch.zeros((w.size(0), d_prime, w.size(2), w.size(3)))
+        rec_weight = w[:, preserve_idx, :, :]
+        if type(op.module) == nn.Linear:
+            rec_weight = rec_weight.squeeze()
+        # no need to provide bias mask for channel pruning
+        rec_mask = torch.ones_like(rec_weight)
+
+        # assign new weight and mask
+        device = op.module.weight.device
+        op.module.weight.data = rec_weight.to(device)
+        op.weight_mask = rec_mask.to(device)
+        if type(op.module) == nn.Conv2d:
+            op.module.in_channels = d_prime
+        else:
+            # Linear
+            op.module.in_features = d_prime
+
+        # export prev layers
+        prev_idx = self.prunable_idx[self.prunable_idx.index(op_idx) - 1]
+        for idx in range(prev_idx, op_idx):
+            m = m_list[idx]
+            if type(m) == nn.Conv2d:  # depthwise
+                m.weight.data = torch.from_numpy(m.weight.data.cpu().numpy()[mask, :, :, :]).to(device)
+                if m.groups == m.in_channels:
+                    m.groups = int(np.sum(mask))
+                m.out_channels = d_prime
+            elif type(m) == nn.BatchNorm2d:
+                m.weight.data = torch.from_numpy(m.weight.data.cpu().numpy()[mask]).to(device)
```

Comment:
it is a little strange here, weight is converted to numpy then back to tensor?

Reply:
Fixed.